pydantic_ai_slim/pydantic_ai/models/openai.py

```diff
@@ -2519,3 +2562,31 @@ def _map_mcp_call(
             provider_name=provider_name,
         ),
     )
+
+
+def _num_tokens_from_messages(
+    messages: list[chat.ChatCompletionMessageParam] | list[responses.ResponseInputItemParam],
+    model: OpenAIModelName,
+) -> int:
+    """Return the number of tokens used by a list of messages."""
+    try:
+        encoding = tiktoken.encoding_for_model(model)
+    except KeyError:
+        encoding = tiktoken.get_encoding('o200k_base')
+
+    if 'gpt-5' in model:
+        tokens_per_message = 3
+        final_primer = 2  # "reverse engineered" based on test cases
+    else:
+        # Adapted from https://cookbook.openai.com/examples/how_to_count_tokens_with_tiktoken#6-counting-tokens-for-chat-completions-api-calls
+        tokens_per_message = 3
+        final_primer = 3  # every reply is primed with <|start|>assistant<|message|>
+
+    num_tokens = 0
+    for message in messages:
+        num_tokens += tokens_per_message
+        for value in message.values():
+            if isinstance(value, str):
+                num_tokens += len(encoding.encode(value))
```

Comment:
Since this (or the `get_encoding` call further up?) could download a large file, but tiktoken is sync not async, we should wrap the call that may do a download in `_utils.run_in_executor` to run it in a thread

Reply:
The methods which download the encoding file are `tiktoken.encoding_for_model` and `tiktoken.get_encoding`. So they would be wrapped with `_utils.run_in_executor` and then awaited? And `_num_tokens_from_messages` would become async?